### PR TITLE
[WebUI] Implement a "Cancel after current iteration" Button

### PR DESCRIPTION
**Pull request by blhook**
## What was the problem/requirement? (What/Why)
Frequently, I wish to cancel the processing of images, but also want the current image to finalize before I do. To work around this, I need to wait until the current one finishes before pressing the cancel. 

## What was the solution? (How)
* Implemented a button that allows to "Cancel after current iteration," which stores a state in the UI that will attempt to cancel the processing after the current image finishes
  * If the button is pressed again, while it is spinning and before the next iteration happens, this will stop the scheduling of the cancel, and behave as if the button was never pressed. 

### Minor
* Added `.yarn` to `.gitignore` as this was an output folder produced from following Frontend's README

#### Major 
* Changed from a standalone button to a context menu next to the original cancel button. Pressing the context menu will give the drop-down option to select which type of cancel method the user prefers, and they can press that button for canceling in the specified type
* Moved states to system state for cross-screen and toggled cancel types management
* Added in distribution for the target yarn version (allowing any version of yarn to compile successfully), and updated the README to ensure `--immutable` is passed for onboarding developers

#### Minor 
* Updated `.gitignore` to ignore specific yarn folders, as specified by their team - https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored

## How were these changes tested?
* `yarn dev` => Server started successfully
* Manual testing on the development server to ensure the button behaved as expected
* `yarn run  build` => Success

#### Revision 1
* Video showing the UI changes in action
https://user-images.githubusercontent.com/89283782/218347722-3a15ce61-2d8c-4c38-b681-e7a3e79dd595.mov

* Images showing the basic UI changes
image
image

#### Revision 2
* Video showing the UI changes in action
https://user-images.githubusercontent.com/89283782/219901217-048d2912-9b61-4415-85fd-9e8fedb00c79.mov

* Images showing the basic UI changes
(Default state) 
image
(Drop-down context menu active) 
image
(Scheduled cancel selected and running)
image
(Scheduled cancel started)
image

## Notes
* Using `SystemState`'s `currentStatus` variable, when the value is `common:statusIterationComplete` is an alternative to this approach (and would be more optimal as it should prevent the next iteration from even starting), but since the names are within the translations, rather than an enum or other type, this method of tracking the current iteration was used instead. 
* `isLoading` on `IAIIconButton` caused the Icon Button to also be disabled, so the current solution works around that with conditionally rendering the icon of the button instead of passing that value. 
* I don't have context on the development expectation for `dist` folder interactions (and couldn't find any documentation outside of the `.gitignore` mentioning that the folder should remain. Let me know if they need to be modified a certain way.

### CODE REVIEW

**Comment by blhook** [comment]
The failed workflow due to the modified `yarn.lock` file:
```
Run yarn install --frozen-lockfile
yarn install v1.22.19
[1/4] Resolving packages...
warning madge > precinct > detective-postcss > postcss-values-parser > flatten@1.0.3: flatten is deprecated in favor of utility frameworks such as lodash.
error Your lockfile needs to be updated, but yarn was run with `--frozen-lockfile`.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
Error: Process completed with exit code 1.
```

However, after reverting the `yarn.lock` file on my system, cleaning the cache, running on the `main` branch at the latest, I am unable to utilize `yarn install --immutable` or `yarn install --frozen-lockfile`: 
```
>yarn install --immutable
➤ YN0070: Migrating from Yarn 1; automatically enabling the compatibility node-modules linker 👍

➤ YN0000: ┌ Resolution step
➤ YN0061: │ flatten@npm:1.0.3 is deprecated: flatten is deprecated in favor of utility frameworks such as lodash.
➤ YN0032: │ fsevents@npm:2.3.2: Implicit dependencies on node-gyp are discouraged
➤ YN0061: │ @npmcli/move-file@npm:2.0.1 is deprecated: This functionality has been moved to @npmcli/fs
stall, which is explicitly forbidden.
➤ YN0000: └ Completed
➤ YN0000: Failed with errors in 11s 856ms
```

Is there a specific version of yarn I should target to have compatibility with the development cycle? 
```
>yarn -v
3.4.1
>node -v
v18.7.0
>npm -v
8.15.0
```

File: invokeai/frontend/src/features/parameters/components/ProcessButtons/CancelButton.tsx
Comment (psychedelicious) [inline invokeai/frontend/src/features/parameters/components/ProcessButtons/CancelButton.tsx:87]: Let's use Chakra's `ButtonSpinner` component here instead of `Spinner` (it is the spinner that is displayed when `isLoading` is used).

File: invokeai/frontend/.yarnrc.yml
Comment (psychedelicious) [inline invokeai/frontend/.yarnrc.yml:1]: I guess this is a yarn v2+ file?

File: invokeai/frontend/.gitignore
Comment (psychedelicious) [inline invokeai/frontend/.gitignore:31]: And this is a yarn v2+ file also?

**Review by psychedelicious — CHANGES_REQUESTED** [review]
Thanks for the PR! I've requested some minor changes.

We're still on yarn v1. That will be the cause of lockfile issues. For reference, this is what I have installed:
```
> yarn -v
1.22.19
> node -v
v18.14.0
> npm -v
9.3.1
```

If you are up for it, I believe this will fix it:
- Revert the changes to `.yarnrc.yml` and `yarn.lock`
- Run `yarn policies set-version 1.22.19` in the yarn package root to set the project's yarn version (doesn't change your global yarn setup)
- Run `yarn`, which I *hope* will auto-magically tell your system to use yarn classic on this project only
- Commit the end result along with other changes

Regarding the dist folder, we are committing the build artifacts to the repo manually at this time. Please do run `yarn build` before committing and include `dist/` in the commit.

**Comment by blessedcoolant** [comment]
Does this need to be another button? Why not a check box of some sort? If its checked, it cancels after the current iteration. Else it cancels then and there?

**Comment by psychedelicious** [comment]
Hmm. It makes sense to always have immediate access to "cancel asap" and "cancel after current iteration".

Two buttons is a bit confusing, but a checkbox or toggle needs a label also. I'm not sure what makes sense from a UX perspective. 

@hipsterusername @netsvetaev 

What do you guys think?

**Comment by netsvetaev** [comment]
> Hmm. It makes sense to always have immediate access to "cancel asap" and "cancel after current iteration".
> 
> Two buttons is a bit confusing, but a checkbox or toggle needs a label also. I'm not sure what makes sense from a UX perspective. 
> 
> @hipsterusername @netsvetaev 
> 
> What do you guys think?

It seems to me that only one cancel button should be left for the generation process (hide invoke and show cancel instead). This way we can use the same layout with two buttons.
Another way is to use some kind of context menu, show it on the right button click/long tap at the usual cancel button and choose there «finish iteration and cancel».

**Comment by hipsterusername** [comment]
> > Hmm. It makes sense to always have immediate access to "cancel asap" and "cancel after current iteration".
> > Two buttons is a bit confusing, but a checkbox or toggle needs a label also. I'm not sure what makes sense from a UX perspective.
> > @hipsterusername @netsvetaev
> > What do you guys think?
> 
> It seems to me that only one cancel button should be left for the generation process (hide invoke and show cancel instead). This way we can use the same layout with two buttons. Another way is to use some kind of context menu, show it on the right button click/long tap at the usual cancel button and choose there «finish iteration and cancel».

This is how I'd imagine this - The cancel button with a context menu/dropdown that allows the user to use the alternative cancellation method. Then, if that method is used, the Cancel button changes to that as the "default", and allows the user to dropdown to change back.

**Comment by blhook** [comment]
For the Yarn problems, I'll try that. I'll also add a note to the README to indicate that usage if it works. 

The context drop-down makea sense to me. I was worried initially about that solution as I can see use-cases where you'd want to cancel quickly, even if you normally utilize the new button (such as UI locks from GPU capped or a slow click on the scheduled cancel immediately after a generation occurs, rather than right before it ends), but they are edge cases to the normal flow.  

I'll see if I can get it done by the end of this weekend. If not, I'll cancel the PR and finish it as I can find time.

**Comment by blhook** [comment]
I was able to get the yarn manager working, although it was slightly different, since `yarn policies ...` couldn't work unless `yarn install` was successful, which it couldn't do so without editing the package manager. `yarn set version ...` did the trick, which also should make it so others can install without worrying about needing to run the command. I'll include that in my next PR revision.

For the button aspect, I had a chance to play with prototyping the suggested route. Following the feel from mail providers that have an scheduled send, the context menu is to the right as a drop-down with a very tiny `background-color-secondary` distinguishing the two button actions:
image

Once clicked, it'll open the menu and provide the cancel types for the users: 
image

Unlike mail providers, pressing the button doesn't activate the cancel of it. It'll only switch the icon for the ability to click that cancel button.  

The context menu will always be available and is never disabled: 
image

Type (and also cancel after iteration) is now saved within the system state context, so it persists as long as that context is held. 

Before I clean it up and send the next PR, @hipsterusername & @netsvetaev & et. al., does this UX match what you were thinking?

---------
Unrelated, but there also appears to be an issue with the `IAIIconButton` hover color not changing appropriately for the existing Invoke/Cancel buttons. It does work with the other `IAIconButton` such as Show Uploads (probably as they are not overriding the initial styling class). If I can debug what is going on by time I do the cleanup, I'll also throw it in the PR. If not, I'll make a new issue for it for tracking.

**Comment by hipsterusername** [comment]
This is excellent. Thanks a ton for the contribution @blhook! 

If you've got interest in some other fun projects, we've got a good list going :)

**Comment by psychedelicious** [comment]
Looks great! Thanks again for your work. 

There is a chakra ui HOC for grouping buttons like you did with the cancel button and drop down: `ButtonGroup`. In case you have grouped them manually with css, I just wanted to direct you to that component in case it makes things simpler. There's no need to change the implementation if you are happy with it. 

`IAIIconButton` is a relatively thin wrapper around `IconButton` and I wonder if some of the `...rest` props are not making it. Some of the `IAI` components are implemented more completely than others.

**Comment by netsvetaev** [comment]
> Before I clean it up and send the next PR, @hipsterusername, @netsvetaev, et. al. does this UX match what you were thinking?

Very nice!

**Comment by blessedcoolant** [comment]
@blhook The new design is better. Ping us when you push a PR. We'll review it and merge. Thanks for your work on this.

**Comment by lstein** [comment]
This PR is consistently failing the Lint test. Could you fix and push?

**Comment by blhook** [comment]
> This PR is consistently failing the Lint test. Could you fix and push?

@lstein , the issue is fixed on my local repo, but the green theme is creating an inconsistent (and unreadable) style while using Chakra's Menu. My available time to work on it is low, but I'm pretty sure I can grind it out by the end of the weekend.
That said, we have a few options:
1. Push PR without green theme changes to ensure workflow passes, but shouldn't be approved/pushed to main
2. Close PR and wait until I finalize the work on my own time, where I can push it after
3. Wait until I get the fix for the green theme in place and push the PR

Let me know which route you'd prefer.

**Comment by psychedelicious** [comment]
We're happy to have this be a marked as a draft until you're ready for a review. I think @lstein may have missed the context where this isn't quite ready yet.

File: invokeai/frontend/.yarnrc.yml
Comment (blhook) [reply invokeai/frontend/.yarnrc.yml]: Yes. The newest revision should clear some of the confusion as it utilizes a specific yarn version to standardize whichever the developer is utilizing. It'll still include this file, but only as much to relate it back to the specific version of yarn the project needs to run without getting into trouble.

File: invokeai/frontend/.gitignore
Comment (blhook) [reply invokeai/frontend/.gitignore]: Yes. Refer to the other comment. This'll still be needed in the commit, but has been fine-tuned to Yarn's recommendations.

**Comment by blhook** [comment]
> This is excellent. Thanks a ton for the contribution @blhook!
> 
> If you've got interest in some other fun projects, we've got a good list going :)

Thanks. I wish I had more free time to play around with fun projects. I may get one or two of my personal big wants in, such as a queuing system and some sort of dynamic prompt handling for variant words/random selection of words come scheduled in at some point. It'd be unfair to promise or commit to anything, though, as I know I can't dedicate the time to actually craft it in a reasonable amount of time. 

---------------------------------

I've created an issue tracking the [hover highlight issue](https://github.com/invoke-ai/InvokeAI/issues/2718). The short-term solution to bypass the problem is simple, but I didn't want to include it within this PR as it has further reaching implications on the code than this PR really warrants.

---------------------------------

@blessedcoolant & @psychedelicious I believe this should be ready for your review now. The green theme handling for the custom styling of the Menu took longer than expected, but it looks significantly better than it defaulted to. It also might be worth removing `theme.ts` since it is a bit deceptive in its current lack of usage, assuming there is no intent to reapply custom themes that way.

**Comment by blessedcoolant** [comment]
Any reason why the yarn stuff is committed with this?

**Comment by blhook** [comment]
> Any reason why the yarn stuff is committed with this?

This was the only way it allowed to build on my local system without directly uninstalling/reinstalling V1. The items committed should allow reproducibility across all different versions of Yarn and be worth adding into the code base to ensure all contributors don't fall into a similar pitfall if they just do a fresh install. 
Can separate out to a different PR if you prefer.

**Comment by blessedcoolant** [comment]
I've pushed a couple of updates. I built a simple menu component called `IAISimpleMenu` .. this should standardize any simple menus we might use. Changed the Cancel menu you added to that. I'll flush out a Full fledged InvokeAI Menu at a later time when we need it.

Also got rid of the CSS. It's all styled inside the component now.

The functionality seems to be as intended. So no problems on that end.

I think this is good to go. Thanks @blhook for the work.

I'll wait for @psychedelicious to review this too before we merge it.

**Review by blessedcoolant — APPROVED** [review]

**Review by blessedcoolant — APPROVED** [review]

**Review by blessedcoolant — APPROVED** [review]

**Review by psychedelicious — APPROVED** [review]
I added `cancelOptions.cancelAfter` to the `redux-persist` blacklist.

This addresses an edge case where generation fails and no exception is caught, and the cancel button will not re-enable itself even after refreshing the page. This can happen if there is an OOM error during generation, as these are caught before they reach the web server.

---

@blhook We'd love more contributions. Please check in with us before putting any time into substantial changes like queuing. We are soon migrating to a new server and execution engine which will handle this. See #1650

**Pull request merged**
